feat: option to use go-templates in custom-fields (#488)
belastingdienst/opr-paas: docs/administrators-guide/capabilities.md

### Notes on templating

- Template takes precendence over default. With custom fields with a configured template defaults have no effect.
  Same goes for validation and required. They jave no effect on templated custom fields.
- Paas takes precedence over templates. When a custom field is configured in the paas, the value can be overruled in the Paas.
  Validation and 'being required' do take effect for these fields.
  **note**: mixing default and required with templates is not advicable.
- Templates return a string, which if it can be parsed as yaml into a map or list,
  will create a multivalue, where the custom field name is suffixed with either the map keys, or list indexes.

### Examples of templating

You can now generate a argocd policy by ranging over the groups in the paas:

!!! example

    ```yml
    apiVersion: cpet.belastingdienst.nl/v1alpha1
    kind: PaasConfig
    metadata:
      name: opr-paas-config
    spec:
      clusterwide_argocd_namespace: paas-capabilities-argocd
      capabilities:
        mycap:
          ApplicationSet: mycap-as
          custom_fields:

>             "argocd-policies":

Reviewer: ```suggestion
            argocd-policies:
```

>               template: |

Reviewer: Please add an example to remove the generated + and trailing line

Author: I changed the block so that |+ is not there anymore, and added a note on why and how it works.

                system:cluster-admins, role:admin
                {{ range $groupName, $group := .Paas.Spec.Groups }}g, {{ $groupName }}, role:admin
                {{end}}
            my-custom-revision:
              validation: '^(main|develop|feature-.*)$'
              default: main
          quotas:
            defaults:
              limits.cpu: "8"
    ```

> You can also enable debugging in the capability when paasconfig has debugging enabled:

Reviewer: Something like: You can reference values from the PaasConfig as well by ....

!!! example

    ```yml
    apiVersion: cpet.belastingdienst.nl/v1alpha1
    kind: PaasConfig
    metadata:
      name: opr-paas-config
    spec:
      clusterwide_argocd_namespace: paas-capabilities-argocd
      capabilities:
        mycap:
          ApplicationSet: mycap-as
          custom_fields:

>             "debug":

Reviewer: ```suggestion
            debug:
```

              template: |

>                 {{ .Config.Debug }}

Reviewer: ```suggestion
                {{ .Config.Spec.Debug }}
```

            my-custom-revision:
              validation: '^(main|develop|feature-.*)$'
              default: main
          quotas:
            defaults:
              limits.cpu: "8"
    ```

You can return a map and create multiple keys (string suffix).

This would create 2 keys:

!!! example

    ```yml
    apiVersion: cpet.belastingdienst.nl/v1alpha1
    kind: PaasConfig
    metadata:
      name: opr-paas-config
    spec:
      clusterwide_argocd_namespace: paas-capabilities-argocd
      capabilities:
        mycap:
          ApplicationSet: mycap-as
          custom_fields:
            "paas_config":
              template: |

>                 debug: {{ .Config.Debug }}

Reviewer: ```suggestion
                debug: {{ .Config.Spec.Debug }}
```

>                 argo: {{ .Config.ArgoEnabled }}

Reviewer: ```suggestion
                argo: {{ .Config.Spec.ArgoEnabled }}
```

            my-custom-revision:
              validation: '^(main|develop|feature-.*)$'
              default: main
          quotas:
            defaults:
              limits.cpu: "8"
    ```

> Like so:

Reviewer: ```suggestion
Which results in the following applicationSet entries:
```

!!! example

    ```yml
    apiVersion: argoproj.io/v1alpha1
    kind: ApplicationSet
    metadata:
      name: mycap-as
      namespace: paas-capabilities-argocd
    spec:
      generators:
        - list:
            elements:
              - paas_config_debug: true
                paas_config_argo: false
      ...
    ```

> You can also return a list and create multiple keys (number suffix).

Reviewer: ```suggestion
You can also specify a list in the .Template spec and create multiple keys (number suffix).
```

This would create 3 keys:

!!! example

    ```yml
    apiVersion: cpet.belastingdienst.nl/v1alpha1
    kind: PaasConfig
    metadata:
      name: opr-paas-config
    spec:
      clusterwide_argocd_namespace: paas-capabilities-argocd
      capabilities:
        mycap:
          ApplicationSet: mycap-as
          custom_fields:
            "paas_config":
              template: |

>                 - {{ .Config.Debug }}

Reviewer: ```suggestion
                - {{ .Config.Spec.Debug }}
```

>                 - {{ .Config.ArgoEnabled }}

Reviewer: ```suggestion
                - {{ .Config.Spec.ArgoEnabled }}
```
